File: lib/rplayer.dart
import 'dart:ffi';
import 'dart:io';

import 'package:ffi/ffi.dart';

import "rplayer_ffi.g.dart";

enum RPlayerStatus { kUnknown, kReadyToPlay, kFailed }

class RPlayer {
  static final _lib = NativeLibrary(Platform.isAndroid
      ? DynamicLibrary.open('librplayer.so')
      : DynamicLibrary.process());

  static RPlayer createFromURL(String url) {
    return RPlayer._(_lib.RPlayer_CreateFromURL(url.toNativeUtf8().cast()));
  }

  RPlayer._(this._id);

  final int _id;

  RPlayerStatus get status {
    return _lib.RPlayer_GetStatus(_id) as RPlayerStatus;
  }

  String get error {
    return _lib.RPlayer_GetError(_id).toString();
  }

  void play() {
    return _lib.RPlayer_Play(_id);
  }

  void pause() {
    return _lib.RPlayer_Pause(_id);
  }

  void dispose() {
    return _lib.RPlayer_Release(_id);
  }
}

File: lib/cpp/src/rplayer.cc
/*
 * Created by Autokaka (qq1909698494@gmail.com) on 2022/05/01.
 */

#include <rplayer/rplayer.h>

RPlayerPtr RPlayer::CreateFromURL(const std::string& url) {
  return std::make_shared<RPlayer>(url);
}

RPlayer::RPlayer(const std::string& url) : decoder_(url) {
  decoder_.delegate = this;
}

void RPlayer::Play() {
  PlayerLike::Play();
  decoder_.Play();
}

void RPlayer::Pause() {
  PlayerLike::Pause();
  decoder_.Pause();
}

void RPlayer::Stop() {
  PlayerLike::Stop();
  decoder_.Stop();
}

// Delegate

void RPlayer::PlayerThrowsError(const std::string& error) {
  Stop();

  std::scoped_lock lock(mutex_);
  status_ = RPlayerStatus::kFailed;
  error_ = error;
}

void RPlayer::PlayerStatusWillChange(RPlayerStatus status) {
  status_ = status;
}

void RPlayer::PlayerStatusDidChange(RPlayerStatus status) {}
